src/System.Net.HttpListener/src/System/Net/Windows/AsyncRequestContext.cs

```diff
@@ -62,19 +91,26 @@ protected override void Dispose(bool disposing)
                 Debug.Assert(!disposing, "AsyncRequestContext::Dispose()|Must call ReleasePins() before calling Dispose().");
                 if (!Environment.HasShutdownStarted || disposing)
                 {
+#if DEBUG
+                    DebugRefCountReleaseNativeOverlapped();
+#endif
                     _boundHandle.FreeNativeOverlapped(_nativeOverlapped);
                 }
             }
+
             base.Dispose(disposing);
         }
 
         internal NativeOverlapped* NativeOverlapped
         {
             get
             {
+#if DEBUG
+                Debug.Assert(Interlocked.Increment(ref _nativeOverlappedUsed) == 1, "NativeOverlapped reused.");
+#endif
```

Comment:
That's true. The current design pattern is brittle and should be changed to something where the NativeOverlapped is guaranteed to be used only once. (The design should follow encapsulation and the principle of single responsibility: NativeOverlapped shouldn't be accessible through a public getter.)